### PR TITLE
Sigrok fwextract kingst la2016 fixup extraction

### PR DESCRIPTION
**Pull request by Planet911**
This fixes extraction of the fpga bitstream for LA2016 (trivial stream name change by vendor broke it, I assume).
This pr also extracts other bitstreams for required for different variations of this device, shows crc32 of extracted files and adds a README generated from the (updated) groff file.

### CODE REVIEW

**Comment by resonantworks** [comment]
@Planet911 thanks so much for this! Tested working perfectly on Linux.

**Comment by Planet911** [comment]
On 2021-05-23 10:23, Matthew Brown wrote:

> @Planet911 [1] thanks so much for this! Tested working perfectly on 
> Linux.
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [2], or unsubscribe 
> [3].

Hi Matthew,

Glad the extraction script works for you, but you might find pulseview 
does not work well with your LA2016. It didn't work at all for me.

For reference the current LA2016 libsigrok driver code is shown here:

https://sigrok.org/gitweb/?p=libsigrok.git;a=tree;f=src/hardware/kingst-la2016;hb=HEAD

Firstly, check the bitstream filename which is at the top of protocol.c, 
it currently looks for kingst-la2016a-fpga.bitstream. However, if you 
have a newer unit, you may need to use kingst-la2016a1-fpga.bitstream 
which was also extracted by the script. The current driver code doesn't 
handle different hardware versions.

If pulseview doesn't work for you, I suggest building libsigrok from 
source and apply my patch from here:

https://patch-diff.githubusercontent.com/raw/sigrokproject/libsigrok/pull/131.patch

My patch handles the different hardware versions (selecting fpga 
firmware as appropriate) and applies some other fixes. LA2016 and LA1016 
now work fine for me in pulseview.

Kevin / Planet911

Links:
------
[1] https://github.com/Planet911
[2] 
https://github.com/sigrokproject/sigrok-util/pull/7#issuecomment-846531825
[3]

**Comment by wsakernel** [comment]
I assume this overrides PR #3, or?

File: firmware/kingst-la/sigrok-fwextract-kingst-la2016 [hunk 1]
```diff
@@ -192,5 +194,31 @@ if __name__ == "__main__":
...
+    saved 5430 bytes to kingst-la-01a2.fw (crc32=720551a9)
+    saved 178362 bytes to kingst-la2016a1-fpga.bitstream (crc32=7cc894fa)
+    saved 178542 bytes to kingst-la2016-fpga.bitstream (crc32=20694ff1)
+    saved 178379 bytes to kingst-la1016a1-fpga.bitstream (crc32=166866be)
+    saved 178151 bytes to kingst-la1016-fpga.bitstream (crc32=7db70001)
+    '''
+
+    # extract all firmware and fpga bitstreams
+    # writer.extract_re(r"fwfpga/(.*)", r"kingst-\1-fpga.bitstream")
+    # writer.extract_re(r"fwusb/fw(.*)", r"kingst-la-\1.fw", decoder=maybe_intel_hex_as_blob)
```
Comment (wsakernel) [inline firmware/kingst-la/sigrok-fwextract-kingst-la2016:212]: If this is all commented out, can't it just go? We have the old code in the repository.

File: firmware/kingst-la/make-readme.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+groff -Tascii -man sigrok-fwextract-kingst-la2016.1 | col -bx > README
+cat README
+
```
Comment (wsakernel) [inline firmware/kingst-la/make-readme.sh:4]: I'd say the README generation should not only be a separate patch but it should also be done for all .1-files in firmware/. So, it will be consistent in the tree. Also, I think the script should be a Makefile (one directory above, of course) probably?

File: firmware/kingst-la/sigrok-fwextract-kingst-la2016.1 [hunk 3]
```diff
@@ -61,4 +63,5 @@ Vendor website
 .RS 4
 .RB http://www.qdkingst.com/download
 .br
-\%MD5 of v3.4.0: ca407133cb83b700983d2b704a4255c2
+\%MD5 of v3.5.0: 812bbd37a16d315a489ca33ea2868a36
+
```
Comment (wsakernel) [inline firmware/kingst-la/sigrok-fwextract-kingst-la2016.1:67]: Add 3.5.1 here? I recall it was mentioned somewhere.

**Review by wsakernel — CHANGES_REQUESTED** [review]

File: firmware/kingst-la/sigrok-fwextract-kingst-la2016 (hunk 1 above)
Comment (Planet911) [reply firmware/kingst-la/sigrok-fwextract-kingst-la2016:212]: If the manufacturer changes the firmware stream names in the resource file, these lines will be useful to discover those new names, and any other firmwares within the resource file. So I recommend to leave this commented code.

File: firmware/kingst-la/make-readme.sh (hunk 2 above)
Comment (Planet911) [reply firmware/kingst-la/make-readme.sh:4]: Fair comment but my makefile foo is very weak :-(
Maybe someone else can help with this.

File: firmware/kingst-la/sigrok-fwextract-kingst-la2016.1 (hunk 3 above)
Comment (Planet911) [reply firmware/kingst-la/sigrok-fwextract-kingst-la2016.1:67]: I was deliberately ignoring minor versions, I see there is 3.5.2 and 3.5.3 now.

File: firmware/kingst-la/sigrok-fwextract-kingst-la2016 (hunk 1 above)
Comment (wsakernel) [reply firmware/kingst-la/sigrok-fwextract-kingst-la2016:212]: I don't fully get it, but I have no strong opinion about it. Maybe a comment explaining why we keep commented out code is useful?

File: firmware/kingst-la/make-readme.sh (hunk 2 above)
Comment (wsakernel) [reply firmware/kingst-la/make-readme.sh:4]: Then I suggest to drop the README creation from this patch. We can discuss if we want READMEs separately. If the conclusion is we want them, then I can create the Makefile for it.

File: firmware/kingst-la/sigrok-fwextract-kingst-la2016.1 (hunk 3 above)
Comment (wsakernel) [reply firmware/kingst-la/sigrok-fwextract-kingst-la2016.1:67]: Why? Because it is too much work? Because I think it would be helpful for users.

Comment (Planet911) [reply firmware/kingst-la/sigrok-fwextract-kingst-la2016.1:67]: I submitted this PR some five months ago, in that time there has been 2 or 3 minor software updates by the vendor. IMHO, the sigrok pull-rate would never keep up with the rate of vendor 'minor' software releases anyway.

Comment (wsakernel) [reply firmware/kingst-la/sigrok-fwextract-kingst-la2016.1:67]: That's true, we won't keep up. But maybe we could add the known versions at the time we update this file. It wouldn't be up-to-date, but it would be less outdated than just major versions. Anyhow, I think it would be nice to have. If it is too much of a burden for you, we can also skip it for now.

**Comment by abraxa** [comment]
Thanks for your work, Planet911! I merged the commits as 03a034f47b9db6bd24035a9ccc0bc01c13217871 and following.
